### ISSUE TITLE
Emit warnings on parameter list in closures after {

### ISSUE THREAD

**Issue by muja** [rust-lang/rust#27300]
# Reproducing steps

This code compiles:

```rust
let p = Some(45).and_then({|x|
    Some(x * 2)
});
```

While this doesn't:

```rust
let p = Some(45).and_then({|x|
    println!("doubling {}", x);
    Some(x * 2)
})
```

The error message is absolutely unintuitive and won't help anyone:

```
<anon>:4:14: 4:15 error: unresolved name `x` [E0425]
<anon>:4         Some(x * 2)
                      ^
error: aborting due to previous error
```

[Playpen link](http://is.gd/vgOyRD)

# Proposed solution
Emit a warning when encountering this fallible syntax.

# Anecdotic background

Having used a lot of Ruby, I had grown way accustomed to the `{ |x| x * 2 }` syntax. Although I had seen the proper Rust syntax a couple of times, it never was emphasized and the significance never stuck. I spent 30+ minutes trying to figure out why a println! statement breaks the build, generating a random error message that has no obvious connections to the actual bug at hand. It shouldn't happen.

**Comment by Xirdus** [comment]
I don't think we should make custom-tailored error messages for people trying to write code in their favorite language and then compile it with rustc. Just like we don't issue any special warning on assigning result of assignment to a variable, we shouldn't issue a warning when someone tries to write Ruby lambdas in Rust. People should learn that even if something looks similar, it doesn't mean that it works the same as in their old language.

**Comment by muja** [comment]
@Xirdus nobody said anything about custom tailored messages and the problem is that the code from Ruby **does** compile in the first place (see the very first code snippet). Changing that, however, would be a breaking change, so instead, I proposed the solution as a warning that actually makes sense! Maybe don't even change the warning, just make a "note: did you mean `|x| { ... }`" ? It would help a lot of people since Rubyists are a target group of the Rust language.

**Comment by killercup** [comment]
> This code compiles (and, in my opinion, it shouldn't):

Just to clarify (for you and anyone who may stumble over this syntax as well and finds this issue):

```rust
let p = Some(45).and_then({|x|
    Some(x * 2)
});
```

is the same as

```rust
let p = Some(45).and_then({
    // This is a block with only one expression.
    // It returns this expression (the closure):
    |x| { Some(x * 2) } // braces optional
});
```

This is totally legitimate and should compile.

Your other example

```rust
let p = Some(45).and_then({|x|
    println!("doubling {}", x);
    Some(x * 2)
})
```

could be written as

```rust
let p = Some(45).and_then({
    |x| { println!("doubling {}", x) }; // braces optional
    Some(x * 2) // returns `Some(i32)`
})
```

I totally understand why rustc's error message is confusing. A solution would be to reformat your code shown in the error and annotating the blocks. (This might be totally confusing in other cases, though!)

**Comment by muja** [comment]
@killercup thanks for the response! So you're saying we need `rustfmt`? I'm all for that and if that fixes the issue, I don't see a problem!

**Comment by killercup** [comment]
@muja you can already use <https://github.com/nrc/rustfmt>! (That isn't integrated into the compiler, though, so the error snippets won't get pretty by themselves.)

**Comment by stevenblenkinsop** [comment]
In the first case, there could plausibly be a hint if you try to use a closure parameter in a subsequent statement, such as

    <anon>:4:14: 4:15 error: unresolved name `x` [E0425]
    <anon>:4         Some(x * 2)
                          ^
    <anon>:2:33: 2:34 note: nearest candidate defined here
    <anon>:2    let p = Some(45).and_then({|x|
                                  ^
    <anon>:3:35: 3:36 note: but goes out of scope here
    <anon>:3        println!("doubling {}", x);
                                              ^
    error: aborting due to previous error

For the second case, plausibly, when the expression for a closure isn't a block, you could have a hint:

    <anon>:2:22: 5:7 error: the trait `core::ops::FnOnce<(_,)>` is not implemented for the type `core::option::Option<_>` [E0277]
    <anon>:2     let p = Some(45).and_then({|x|
    <anon>:3         println!("doubling {}", x);
    <anon>:4         Some(100 * 2)
    <anon>:5     });
    <anon>:2:22: 5:7 help: see the detailed explanation for E0277
    <anon>:4:9: 4:22 note: the expression here
    <anon>:4         Some(100 * 2)
                    ^~~~~~~~~~~~~
     <anon>:2:32: 3:35 note: may have been intended to be part of closure expression here:
    <anon>:2     let p = Some(45).and_then({|x|
    <anon>:3         println!("doubling {}", x);
    <anon>:3:34: 3:35 note: but closure expression ends here
    <anon>:3         println!("doubling {}", x);
                                             ^
    error: aborting due to 2 previous errors

**Comment by withoutboats** [comment]
Note that if the unbound variable error hadn't arisen (e.g. because `x` was defined in the outer scope), this code would have not compiled because the block evaluates to `Some(x * 2)`, which does not implement `FnOnce(()) -> Option<U>`. You just almost definitely never want to pass a block to any parameter that expects a function, so there's that.

There are a couple of different ways rustc could generate a more helpful error message here. If an unbound variable is found in an identifier in the same scope as a previous closure statement that uses that identifier, it could point out that the statement is not inside the closure. If a closure is the first statement of a block and not bound with a `let` statement, it could remind you of Rust's closure syntax. Both of these would be reasonable, and helpful to users confused about closure syntax, though I don't know how easy either of them is.

**Comment by estebank** [comment]
Given #47763, overly verbose, but educative output actually pointing at the problem:

```
warning: a closure's body is not determined by its enclosing block
  --> $DIR/block-enclosed-closure.rs:14:33
   |
14 |     let _p = Some(45).and_then({|x|
   |                                 ^^^ this closure's body is not determined by its enclosing block
   |
note: this is the closure's block...
  --> $DIR/block-enclosed-closure.rs:14:33
   |
14 |       let _p = Some(45).and_then({|x|
   |  _________________________________^
17 | |         println!("hi");
   | |______________________^
note: ...while this enclosing block...
  --> $DIR/block-enclosed-closure.rs:14:32
   |
14 |       let _p = Some(45).and_then({|x|
   |  ________________________________^
17 | |         println!("hi");
18 | |         Some(x * 2)
19 | |     });
   | |_____^
note: ...implicitely returns
  --> $DIR/block-enclosed-closure.rs:18:9
   |
18 |         Some(x * 2)
   |         ^^^^^^^^^^^
help: you should open the block *after* the closure's argument list
   |
14 |     let _p = Some(45).and_then(|x| {
   |                                ^^^^^

error[E0425]: cannot find value `x` in this scope
  --> $DIR/block-enclosed-closure.rs:18:14
   |
18 |         Some(x * 2)
   |              ^ not found in this scope

error[E0277]: the trait bound `std::option::Option<_>: std::ops::FnOnce<({integer},)>` is not satisfied
  --> $DIR/block-enclosed-closure.rs:14:23
   |
14 |     let _p = Some(45).and_then({|x|
   |                       ^^^^^^^^ the trait `std::ops::FnOnce<({integer},)>` is not implemented for `std::option::Option<_>`

error: aborting due to 2 previous errors
```

**Comment by estebank** [comment]
For the second case where it is a literal, we emit:

```
error[E0277]: expected a `FnOnce({integer})` closure, found `Option<{integer}>`
    --> src/main.rs:2:31
     |
   2 |       let p = Some(45).and_then({|x|
     |  ______________________--------_^
     | |                      |
     | |                      required by a bound introduced by this call
   3 | |         println!("doubling {}", x);
   4 | |         Some(90)
     | |         -------- this tail expression is of type `Option<{integer}>`
   5 | |     });
     | |_____^ expected an `FnOnce({integer})` closure, found `Option<{integer}>`
     |
     = help: the trait `FnOnce({integer})` is not implemented for `Option<{integer}>`
note: required by a bound in `Option::<T>::and_then`
    --> /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:1537:12
     |
1535 |     pub const fn and_then<U, F>(self, f: F) -> Option<U>
     |                  -------- required by a bound in this associated function
1536 |     where
1537 |         F: [const] FnOnce(T) -> Option<U> + [const] Destruct,
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Option::<T>::and_then`
help: you might have meant to create the closure instead of a block
     |
   2 |     let p = Some(45).and_then(|_| {|x|
     |                               +++
```

Where the output is too verbose, but we provide a suggestion.

For the case where we use the desired argument, we still don't handle it at all:

```
error[E0425]: cannot find value `x` in this scope
 --> src/main.rs:4:14
  |
4 |         Some(x * 2)
  |              ^ not found in this scope
```
